[MRG + 2] FIX LogisticRegressionCV to correctly handle string labels

raghavrv opened the pull request:
Fixes #5868

In master - 

``` python
>>> import numpy as np
>>> from sklearn.linear_model import LogisticRegression, LogisticRegressionCV

>>> X = np.arange(12)[:, np.newaxis]
>>> y = ['a',] * 4 + ['b',] * 4 + ['c',] * 4

>>> LogisticRegressionCV(solver='lbfgs', multi_class='multinomial').fit(X, y).predict(X)
ValueError: could not convert string to float: 'a'
```

In this branch -

``` python
>>> LogisticRegressionCV(solver='lbfgs', multi_class='multinomial').fit(X, y).predict(X)
array(['a', 'a', 'a', 'a', 'b', 'b', 'b', 'b', 'c', 'c', 'c', 'c'], dtype='<U1')
```

@agramfort @GaelVaroquaux

MechCoder commented on sklearn/linear_model/logistic.py:
```diff
@@ -1525,9 +1525,6 @@ def fit(self, X, y, sample_weight=None):
         folds = list(cv.split(X, y))
 
-        self._enc = LabelEncoder()
-        self._enc.fit(y)
```
Surely self._enc must be used somewhere?

rvraghav93 replied:
It was added here - https://github.com/scikit-learn/scikit-learn/commit/67585f6e6f1ee5912fb520befe241fa74db8c37d, but not used anywhere... I am unable to understand why though...

MechCoder commented:
Instead of this can we encode `y` once in the fit time of `LogisticRegressionCV` and pass the encoded y to every split.

If you did this you would need to reconstruct the `class_weights` if they are in a dict format, such that they use the encoded class labels. Other than that I don't foresee any problem.

Also you can remove the `check_classification_targets` to before the type of y is checked.

MechCoder commented:
Would also be a good time to clean up some parts of the code, especially the encoding logic, dtype checks etc

amueller commented:
hm this is still a bug, right?

jnothman commented on sklearn/linear_model/logistic.py:
```diff
@@ -898,7 +897,8 @@ def _log_reg_scoring_path(X, y, train, test, pos_class=None, Cs=10,
 
     # To deal with object dtypes, we need to convert into an array of floats.
-    y_test = check_array(y_test, dtype=np.float64, ensure_2d=False)
+    y_test = check_array(LabelEncoder().fit_transform(y_test),
```
Surely `y_test` needs to use the same encoder as was used for training.

raghavrv commented:
@MechCoder @jnothman @amueller Could you take a look at this now?

TomDLT commented:
What about just adding label encoding after `_check_solver_option` in `_log_reg_scoring_path` ?
```python
le = LabelEncoder()
y = le.fit_transform(y)
```

TomDLT commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -398,6 +399,38 @@ def test_logistic_cv():
+    y = np.array(y) - 1
+    # Test for string labels
+    lr = LogisticRegression(solver='lbfgs', multi_class='multinomial')
+    lr_cv = LogisticRegression(solver='lbfgs', multi_class='multinomial')
```
you probably wanted to use `LogisticRegressionCV` here

jnothman commented on sklearn/linear_model/logistic.py:
```diff
@@ -1557,32 +1554,34 @@ def fit(self, X, y, sample_weight=None):
-        check_consistent_length(X, y)
+        if y.dtype.kind in ('S', 'U'):
+            # Encode for string labels
+            self._label_encoder = LabelEncoder().fit(y)
```
Why can't you do encoding in all cases and assume that `LabelEncoder` handles efficiency issues?

raghavrv replied:
thanks for the comment I did so in the recent commit...

jnothman commented on sklearn/linear_model/logistic.py:
```diff
@@ -1557,32 +1554,34 @@ def fit(self, X, y, sample_weight=None):
+                for new_key, old_key in zip(new_keys, old_keys):
+                    self.class_weight[new_key] = self.class_weight[old_key]
+        else:
+            self.classes_enc_ = self.classes_
```
Does this work when classes are [1,2,3], not [0,1,2]?

raghavrv replied:
I'm encoding for all cases now like you suggested. And there is a test for this usecase.

raghavrv committed:
TST if LogisticRegressionCV handles string labels properly

raghavrv committed:
TST Add a test with class_weight dict

raghavrv commented:
Arghh. This is tougher to solve than I thought. After encoding the string labels at the fit of `LogisticRegressionCV`, the `logistic_regression_path` helper again encodes the data and this causes a test failure I think... I'll have a deeper look at it soon...

raghavrv committed:
ENH Encode y and class_weight dict

raghavrv commented:
Now it should pass all the tests... Gentle ping @TomDLT @MechCoder for reviews too...

TomDLT commented:
Your variables names are quite confusing:
`enc_labels, iter_labels, enc_lbl`
`cls_labels, iter_classes, cls_lbl`

what about:
`encoded_labels, iter_encoded_labels, encoded_label`
`classes_labels, iter_classes_labels, classes_label`
or
`enc_labels, iter_enc_labels, enc_label`
`cls_labels, iter_cls_labels, cls_label`

TomDLT commented:
Otherwise LGTM

raghavrv commented:
For `multiclass='multinomial'` there is a hack that requires the `iter_encoded_lables` to be different from the `encoded_labels`. But I've modified the names as you suggested. Thanks for taking a look at this!

Having addressed your comment I'm adding your +1...

raghavrv committed:
Better variable names

TomDLT commented on sklearn/linear_model/logistic.py:
```diff
@@ -1680,14 +1680,14 @@ def fit(self, X, y, sample_weight=None):
 
-        for index, (cls_lbl, enc_lbl) in enumerate(
-                zip(iter_classes, iter_labels)):
+        for index, (casses_label, encoded_label) in enumerate(
```
casses -> classes

TomDLT commented on sklearn/linear_model/logistic.py:
```diff
@@ -1682,10 +1680,14 @@ def fit(self, X, y, sample_weight=None):
             coefs_paths = multi_coefs_paths
 
-        for index, label in enumerate(iter_labels):
+        for index, (casses_label, encoded_label) in enumerate(
```
casses -> classes

raghavrv committed:
TYPO casses --> classes

raghavrv commented:
@TomDLT Thanks for the catch. Have fixed!

amueller reviewed (commented):
huh I just realized I had pending comments. Still not used to the new github interface :-/

raghavrv commented on sklearn/linear_model/logistic.py:
```diff
@@ -1561,64 +1556,67 @@ def fit(self, X, y, sample_weight=None):
+            new_keys = label_encoder.transform(old_keys)
+            # Don't modify the original class_weight dict.
+            class_weight = dict()
+            for new_key, old_key in zip(new_keys, old_keys):
```
Ah now python 2.6 is removed we can have dict comprehension right? Cool!

raghavrv commented on sklearn/linear_model/logistic.py:
```diff
@@ -1561,64 +1556,67 @@ def fit(self, X, y, sample_weight=None):
             max_squared_sum = None
 
-        check_classification_targets(y)
+        # Encode for string labels
```
`enc_labels` (now `encoded_labels`) refers to the encoded labels that are being considered. (by being considered I mean, the ones that are finally used to train the model, as for binary problems the higher label alone is retained hence differs from `range(len(classes))`).

Ref [this line](https://github.com/scikit-learn/scikit-learn/pull/5874/files#diff-1452d97adf1a0b794930e405aa58d64bR1598)

raghavrv replied:
`classes_labels` was done to match the naming of `encoded_labels`. I'll refactor it back to `classes` if it seems clearer...

raghavrv committed:
FIX Use dict comprehension; classes_labels --> classes

amueller commented:
hm we're still running 2.6 ci...

raghavrv committed:
Don't use dict comprehension

raghavrv commented:
> hm we're still running 2.6 ci...

Oh yea. I don't know why I keep assuming we are removing support for it... Anyway have reverted the dict comprehension fix...

amueller commented:
well we are dropping support. Though I was stupid because this will be backported to 0.18.1 which still supports Python 2.6. So sorry for that....

raghavrv commented:
No problem, I was also confused :) Any more reviews from your side?

raghavrv commented:
Also p{ing|ester} @TomDLT ;)

raghavrv commented:
@agramfort Can I haz a review? :)

agramfort commented:
no time.

@raghavrv @TomDLT or @jnothman are enough as 3 core devs :)

amueller commented:
why do we need a ``LabelEncoder`` in ``LogisticRegressionCV`` at all, when ``logistic_regression_path`` can handle classes internally?

amueller commented:
I guess we can merge this as a bug-fix. The fit method seems to be getting too long and hard to grasp. The variables mean different things in the binary, ovr and multinomial case. But maybe this is not the time for a clean-up. I'm not sure.

raghavrv commented:
> why do we need a `LabelEncoder` in `LogisticRegressionCV` at all, when `logistic_regression_path` can handle classes internally?

Because `log_reg_scoring_path` doesn't handle string labels... I understand it adds quite a few lines, but I can't think of a better way than cleanly encoding all the labels / `class_weight` dict beforehand... :(

amueller commented:
ah. hm well this is fine with me. @jnothman are your concerns addressed? This has +1 from me and @TomDLT, I think.

raghavrv commented:
Would have been nice if github introduced "Ping Reviewer" feature so I can pester Joel/Andy more efficiently in one click :p

jnothman commented:
Browser extension?

On 5 November 2016 at 01:25, Raghav RV <notifications@github.com> wrote:

> Would have been nice if github introduced "Ping Reviewer" feature so I can
> pester Joel/Andy more efficiently in one click :p
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5874#issuecomment-258445056>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz641b5DiHk0USbGYnzQKnC_Gok65Jks5q60BIgaJpZM4Gkt-Q>
> .
>

jnothman commented on sklearn/linear_model/logistic.py:
```diff
@@ -1561,64 +1556,64 @@ def fit(self, X, y, sample_weight=None):
-                DataConversionWarning)
-            y = np.ravel(y)
+        encoded_labels = label_encoder.transform(label_encoder.classes_)
+        classes = self.classes_  # The original class labels
```
but wasn't this set two lines ago?

jnothman commented on sklearn/linear_model/logistic.py:
```diff
@@ -1561,64 +1556,64 @@ def fit(self, X, y, sample_weight=None):
+            iter_encoded_labels = encoded_labels
+            iter_classes = classes
 
-        if self.class_weight and not(isinstance(self.class_weight, dict) or
```
I think clarity could be improved by moving this `class_weight` validation higher up in the function, thus creating more cohesion for uses of `iter_encoded_labels`, etc.

jnothman commented on sklearn/linear_model/logistic.py:
```diff
@@ -1561,64 +1556,64 @@ def fit(self, X, y, sample_weight=None):
+            all_encoded_labels = np.arange(len(self.classes_))
+            class_weight = compute_class_weight(class_weight,
+                                                all_encoded_labels, y)
+            class_weight = dict(zip(all_encoded_labels, class_weight))
```
You could do this with `dict(enumerate(class_weight))` and get rid of `all_encoded_labels` which I think adds confusion to the reader.

raghavrv replied:
But for `compute_class_weight` you need `all_encoded_labels` right? We don't have that information anywhere else as `encoded_labels` is modified a few lines before...

jnothman reviewed (changes requested):
minor things, otherwise LGTM

raghavrv committed:
MNT reorder validation to improve clarity

raghavrv committed:
Use enumerate

raghavrv commented on sklearn/linear_model/logistic.py:
```diff
@@ -1561,64 +1556,65 @@ def fit(self, X, y, sample_weight=None):
-            class_weight = self.class_weight
+        if class_weight in ("auto", "balanced"):
+            class_weight = compute_class_weight(class_weight,
+                                                np.arange(len(self.classes_)),
```
@jnothman does this seem okay? I could use `np.unique(y)`, but this seemed faster...

jnothman replied:
Sure this is fine. I don't get why you'd choose `np.unique(y)` in this context.

raghavrv committed:
BUGFIX

raghavrv committed:
zip was not needed

jnothman commented:
LGTM, merging

jnothman commented:
Nah, not merging. I'll let you add a what's new and merge yourself.

amueller commented:
@raghavrv do you want to do the merge and whatsnew? I want to release asap

raghavrv commented:
Sure sorry for the delay. In 15 mins...

raghavrv committed:
Add whatsnew

raghavrv commented:
Merging once CIs pass...

raghavrv commented:
Thanks @jnothman @amueller @TomDLT @MechCoder for the reviews!

amueller commented:
thanks for the pr :)

jnothman commented:
We seem to be getting test failures from this on PRs, such as at https://ci.appveyor.com/project/sklearn-ci/scikit-learn/build/1.0.10296/job/qol1vrsk30ycsopl

raghavrv commented:
Which PR is this? I saw a similar error but because of an incorrect updation of master. (Which got resolved after fixing that...)

jnothman commented:
https://github.com/scikit-learn/scikit-learn/pull/7838 The commit history
looks clean.

On 21 November 2016 at 09:33, Raghav RV <notifications@github.com> wrote:

> Which PR is this? I saw a similar error but because of an incorrect
> updation of master. (Which got resolved after fixing that...)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5874#issuecomment-261811383>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz68dXJJvBChav9vNhG-Dby-LREw8Oks5rAMrAgaJpZM4Gkt-Q>
> .
>

raghavrv commented:
Okay. Thanks for the notice. I'll look into it tomorrow...
